crates/ide-completion/src/context.rs

```diff
@@ -809,12 +826,76 @@ impl<'a> CompletionContext<'a> {
             qualifier_ctx,
             locals,
             depth_from_crate_root,
+            exclude_flyimport_traits,
+            exclude_traits,
             complete_semicolon,
         };
         Some((ctx, analysis))
     }
 }
 
+fn resolve_exclude_traits_list(db: &RootDatabase, traits: &[String]) -> FxHashSet<hir::Trait> {
```

Comment:
This should be doable in a more simple manner using semantics, `Semantics::scope` (we likely already have that computed at this point) and then expose a method to that that either resolves a string path or a mod path using similar logic of `doc_modpath_from_str` function in `crates\hir\src\attrs.rs`

Reply:
Or actually, do what the snippet stuff is doing by turning the strings into green nodes https://github.com/rust-lang/rust-analyzer/blob/2f55a9155211a9c4a7e061ed4c01bacfa2ebc3d4/crates/ide-completion/src/snippet.rs#L197-L222

it already is doing something like this for paths, creating fake ast nodes for them and then resolving those via the scope at the use site

Reply:
I somewhat dislike this approach since it means the names are not universal: they can be shadowed and will be impacted by dependency renames. This also makes caching them harder, if we will ever want to do that. Not strongly opposed, though.

Reply:
Hmm, well, looking at the logic here though this is also not quite right though. The `display_name` of a crate shouldn't be used for name resolution given the name of a crate really is the name the dependency edge gives it

Reply:
> I somewhat dislike this approach since it means the names are not universal: they can be shadowed and will be impacted by dependency renames. This also makes caching them harder, if we will ever want to do that. Not strongly opposed, though.

Also I think that's the approach we want, the resolution has to be dependent on the scope, crates get their names from the dependency edge (I just realized I am repeating part of my previous comment but whatever), their display name is irrelevant.